core/directory.go

```diff
@@ -164,6 +168,13 @@ func (dir *Directory) Entries(ctx context.Context, gw bkgw.Client, src string) (
 	if err != nil {
 		return nil, err
 	}
+	// empty directory, i.e. llb.Scratch()
```

Comment:
@vito PTAL, I think this fixes #4130 (as well as this PR)

- Copied the original check logic for `llb.Scratch()` above, however check for `res.SingleRef()` returning `nil` (rather than checking `payload.LLB == nil`). AFAIK `gw.Solve() + SingleRef()` will always return `nil` for a `scratch` **result**
- Removed the original scratch check above for repetition. Solving `llb.Scratch()` (or `nil`) will result to a scratch result, so the edge case is already captured (the only downside is we lose a bit of time trying to Solve() scratch, but that's ok)
- Not sure what edge cases `path.Clean(src) == "."` is supposed to catch, but that didn't work for `Directory("")`. So I added an explicit `/` check to make this work for `scratch.Directory("")`, `scratch.Directory("/")`
- Added dir Mode to `Stat` of scratch. `container.Directory()` has a `if !info.IsDir() { return nil, fmt.Errorf("path %s is a file, not a directory", dirPath) }` check, which was failing

/cc @sipsma

Reply:
Nice, `if ref == nil` feels a lot more sturdy. LGTM!